src/Microsoft.Health.Fhir.TaskManagement/TaskHosting.cs

```diff
@@ -0,0 +1,201 @@
+
+        public short MaxRunningTaskCount { get; set; } = Constants.DefaultMaxRunningTaskCount;
+
+        public short MaxRetryCount { get; set; } = Constants.DefaultMaxRetryCount;
+
+        public int TaskHeartbeatTimeoutThresholdInSeconds { get; set; } = Constants.DefaultTaskHeartbeatTimeoutThresholdInSeconds;
+
+        public int TaskHeartbeatIntervalInSeconds { get; set; } = Constants.DefaultTaskHeartbeatIntervalInSeconds;
+
+        public async Task StartAsync(CancellationTokenSource cancellationToken)
+        {
+            CancellationTokenSource keepAliveCancellationToken = new CancellationTokenSource();
+            Task keepAliveTask = KeepAliveTasksAsync(keepAliveCancellationToken.Token);
+
+            await PullAndProcessTasksAsync(cancellationToken.Token);
+
+            keepAliveCancellationToken.Cancel();
+            await keepAliveTask;
+        }
+
+        private async Task PullAndProcessTasksAsync(CancellationToken cancellationToken)
+        {
+            List<Task> runningTasks = new List<Task>();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                Task intervalDelayTask = Task.Delay(TimeSpan.FromSeconds(PollingFrequencyInSeconds));
+
+                if (runningTasks.Count >= MaxRunningTaskCount)
+                {
+                    _ = await Task.WhenAny(runningTasks.ToArray());
+                    runningTasks.RemoveAll(t => t.IsCompleted);
+                }
+
+                IReadOnlyCollection<TaskInfo> nextTasks = null;
+                try
+                {
+                    nextTasks = await _consumer.GetNextMessagesAsync((short)(MaxRunningTaskCount - runningTasks.Count), TaskHeartbeatTimeoutThresholdInSeconds, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Failed to pull new tasks.");
+                }
+
+                if (nextTasks != null && nextTasks.Count > 0)
+                {
+                    foreach (TaskInfo taskInfo in nextTasks)
+                    {
+                        runningTasks.Add(ExecuteTaskAsync(taskInfo, cancellationToken));
+                    }
+                }
+
+                await intervalDelayTask;
+            }
+
+            try
+            {
+                Task.WaitAll(runningTasks.ToArray());
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Task failed to execute");
+            }
+        }
+
+        private async Task ExecuteTaskAsync(TaskInfo taskInfo, CancellationToken cancellationToken)
+        {
+            using ITask task = _taskFactory.Create(taskInfo);
+            task.RunId = task.RunId ?? taskInfo.RunId;
+
+            if (task == null)
+            {
+                _logger.LogWarning("Not supported task type: {0}", taskInfo.TaskTypeId);
```

Comment:
nit: Use structured logging variable
```suggestion
                _logger.LogWarning("Not supported task type: {taskTypeId}", taskInfo.TaskTypeId);
```

Reply:
thanks for the suggestion, updated.